src/internal/observable/forkJoin.ts

```diff
@@ -124,51 +126,40 @@ export function forkJoin<T>(sourcesObject: T): Observable<{ [K in keyof T]: Obse
  */
 export function forkJoin(...args: any[]): Observable<any> {
   const resultSelector = popResultSelector(args);
-
   const { args: sources, keys } = argsArgArrayOrObject(args);
-
-  if (resultSelector) {
-    // deprecated path.
-    return forkJoinInternal(sources, keys).pipe(map((values: any[]) => resultSelector!(...values)));
-  }
-
-  return forkJoinInternal(sources, keys);
-}
-
-function forkJoinInternal(sources: ObservableInput<any>[], keys: string[] | null): Observable<any> {
-  return new Observable((subscriber) => {
-    const len = sources.length;
-    if (len === 0) {
+  const result = new Observable((subscriber) => {
+    const { length } = sources;
+    if (!length) {
       subscriber.complete();
       return;
     }
-    const values = new Array(len);
-    let completed = 0;
-    let emitted = 0;
-    for (let sourceIndex = 0; sourceIndex < len; sourceIndex++) {
-      const source = innerFrom(sources[sourceIndex]);
+    const values = new Array(length);
+    let remainingCompletions = length;
+    let remainingEmissions = length;
```

Comment:
Changed these to a "countdown" instead of a counter. It makes the checks below a little more terse, and I think it's easier to reason about if you were to step into the code "Oh, there's 1 left" instead of having to do (very simple) math in your head to figure out how many more need to complete.